client/state/data-layer/wpcom/sites/rewind/index.js

```diff
@@ -27,17 +33,75 @@ const fetchRewindState = action =>
 		action
 	);
 
+/**
+ * Re-requests the activity log entries for all
+ * previously-sent queries in application state
+ *
+ * This function is highly coupled to the details
+ * of Activity Log and Rewind state!
+ *
+ * It will not grab events we haven't already
+ * gotten. Instead it will reissue the same reqeusts
+ * as previously made so that we get updates on
+ * things like the `isDiscarded` status of events.
+ *
+ * @param {Number} siteId the site being queried
+ * @param {Object} rewind Rewind state for the site
+ * @returns {function} thunk action maybe dispatching requests
+ */
+const refreshActivityLogAfterRewind = ( siteId, rewind ) => ( dispatch, getState ) => {
+	if ( ! rewind || rewind.status === 'running' ) {
```

Comment:
(because you asked for thoughts)

how often does `REWIND_STATE_REQUEST` get dispatched? It seems like there is room to make it more elegant based on how many early return checks are necessary.

In an ideal world, `refreshActivityLogAfterRewind` feels like it should be the response to a `REWIND_COMPLETE` action type as opposed a success handler to `REWIND_STATE_REQUEST`.

Reply:
> it should be the response

either that or I should put this logic in the actual activity log data handler file and have it listen for the returning `REWIND_STATE_REQUEST` network request and save the additional action type.

> how often does REWIND_STATE_REQUEST get dispatched? It seems like there is room to make it more elegant based on how many early return checks are necessary.

the Rewind state governs the entire UI in on sense. we poll for it contextually based on whether it indicates that a process is currently running. that it, it's a declarative approach replacing a number of imperative calls we have elsewhere (see #21855).

the logic in this function however I think is pretty minimal:
 - if we are examine a state without a running process
 - if the last state had a running process
 - if we have already loaded activity log events
 - then re-fetch all of the events

the early aborts are mainly to prevent performing computation we know we don't need.